### PR TITLE
Friends dataset

### PR DESCRIPTION
**Pull request by chiehminwei**
**Patch description**
This patch adds a new task `friends` for generating multiparty conversation training examples from the [Friends Corpus on ConvoKit](https://convokit.cornell.edu/documentation/friends.html). 

Example Training Sample:

```
Text:
Person A: I like pizza.
Person B: Me too.
Person C: Could someone order pizza.

Label:
	I will do it
```

The teacher supports 3 command line arguments, `--character`, `--use-silence-token`, and `--use-start-token`.

The `--character` option specifies which speaker labels to train on. For example, if 'Rachel Green' is specified, then only sentences uttered by Rachel Green will generate a training sample.

The choices for `character` are the main characters from the show, or an `All` option specifying all 6 main characters, excluding supporting cast, should generate training samples. The `All` option is the default option.

```
[
  'All',
  'Rachel Green',
  'Monica Geller',
  'Phoebe Buffay',
  'Joey Tribbiani',
  'Chandler Bing',
  'Ross Geller',
]
```

The `--uses-silence-token` (default: `True`) flag indicates whether a training sample should be generated with a <SILENCE> token when it's not the chosen character's turn to speak. For example, the earlier pizza conversation would generate the following training samples:

```
# Chosen Character = Person C

Text:
Person A: I like pizza.

Label:
    __SILENT__

===
Text:
Person A: I like pizza.
Person B: Me too.

Label:
      Could someone order pizza.
```

The `--uses-start-token` (default: `False`) flag indicates whether a <START> token should be included in the beginning of a conversation. For example, the earlier pizza conversation would generate the following training samples:

```
# --use_start_token = True
# Chosen Character = Person A

Text:
__Start__

Label:
  I like pizza.
```
Whereas when `--use-start-token = False`, the first sentence in a conversation would be skipped and not generate a training sample.

There are also two flags `--start-token` and `--silence-token` that you can use to specify what special symbols you want to use to represent these tokens. By default, they are `__START__` and `__SILENCE__`.

**Testing steps**

```
pytest parlai/tasks/myteacher/test.py
```

Expected Output:

```
============================================================= test session starts ==============================================================
test.py ...
```

**Other information**

To see the examples generated, run:

```
parlai display_data --task friends
```

### CODE REVIEW

**Comment by chiehminwei** [comment]
@mojtaba-komeili

File: parlai/tasks/friends/agents.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+from parlai.core.teachers import DialogTeacher
+from parlai.core.params import ParlaiParser
+from .build import build
+
+import copy
+import os
+import json
+
+
+def _path(opt, *additions):
```
Comment (mojtaba-komeili) [inline parlai/tasks/friends/agents.py:18]: Do you use the additional args? If not why keeping them?

File: parlai/tasks/friends/agents.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    return os.path.join(
+        opt['datapath'], 'Friends', 'friends-corpus/utterances.jsonl', *additions
+    )
+
+
+class DefaultTeacher(DialogTeacher):
+    START_TOKEN = '<START>'
```
Comment (mojtaba-komeili) [inline parlai/tasks/friends/agents.py:25]: For these constants, maybe it makes sense to set them via flags, that way they can be consistent when we are mixing them with other datasets that have similar tokens, See [this](https://github.com/facebookresearch/ParlAI/blob/7660197631c06ee41c0ce25c5de37a74b11ecb77/projects/blenderbot2/agents/blenderbot2.py#L279-L284).

Also, in most of our recent projects we prefer a different style for the special tokens. See the silence token [here](https://github.com/fairinternal/ParlAI-Internal/blob/1ed6496e549944cf02017930a63baa7555f9055a/projects/blenderbot3/tasks.py#L121).

File: parlai/tasks/friends/agents.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        with open(datafile, 'r') as json_file:
+            for json_str in json_file:
+                utterance = json.loads(json_str)
+
+                text = utterance['text']
+                speaker = utterance['speaker']
+                conversation_id = utterance['conversation_id']
+
+                if conversation_id not in conversations:
```
Comment (mojtaba-komeili) [inline parlai/tasks/friends/agents.py:56]: Checkout python's defaultdict for simpler implantation here.

File: parlai/tasks/friends/agents.py (hunk 1 above)
Comment (chiehminwei) [reply parlai/tasks/friends/agents.py:18]: Ahh good catch I'll remove them

**Review by mojtaba-komeili — CHANGES_REQUESTED** [review]
Thanks Jimmy, it looks good. Just a few comments to make it better:
1. try our [autoformat.sh](https://github.com/facebookresearch/ParlAI/blob/main/autoformat.sh) script for fixing the lint errors.
2. Make sure tests are passing. There are some failing now.
3. Add convokit to the installation requirements (maybe that is what causing test failures)
4. It's alright to have a single teacher, but maybe add a couple more for convenience use. They may be simple extensions of the current one with minimal changes. For example a teacher for All and another one that is a single charter POV.

**Comment by mojtaba-komeili** [comment]
One more thing, could you add the list of existing character as another entry to the message? It could be a comma separated list of names.

**Comment by chiehminwei** [comment]
Sounds great, all done! See updated PR. 

Used `autoformat.sh`, got around ConvoKit, added multiple convenience Teacher classes and added corresponding tests, and added cli option to specify the list of characters. All tests passed except the cleaninstall one unrelated to this PR.

File: parlai/core/build_data.py [hunk 4]
```diff
@@ -380,6 +380,8 @@ def _unzip(path, fname, delete=True):
                 PathManager.mkdirs(outpath)
                 continue
             logging.debug(f"Extracting to {outpath}")
+            if '__MACOSX' in member:
```
Comment (mojtaba-komeili) [inline parlai/core/build_data.py:383]: It logically makes sense to submit this as its own patch. Please remove it from this PR and add is as a separate one.

File: parlai/tasks/friends/build.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    with jsonlines.open(test_datafile, mode='w') as writer:
+        for conversation_id in test:
+            for utterance in conversations[conversation_id]:
+                writer.write(utterance)
+
+
+def build(opt):
+    dpath = os.path.join(opt['datapath'], 'Friends')
+    version = '1.02'
```
Comment (mojtaba-komeili) [inline parlai/tasks/friends/build.py:80]: You can change this to 1 since it was never published to the public code. But remove your local files to force a rebuild.

**Comment by mojtaba-komeili** [comment]
One more thing that forgot asking you to add is sample screenshots that shows the output when you run `parlai display_data -t <your task name>`. Just a couple of teachers would be enough.

**Comment by mojtaba-komeili** [comment]
> > 
> 
> This this what you mean? https://github.com/facebookresearch/ParlAI/pull/4568/files#r891537386

(For the record keeping) we chatted about this in-person.

**Comment by chiehminwei** [comment]
## Screenshots

`parlai display_data --task friends`
image

`parlai display_data --task friends:rachel`
image

parlai display_data --task friends:rachel --use-silence-token=False
image

`parlai display_data --task friends:monica`
image

**Review by mojtaba-komeili — APPROVED** [review]
Looks good thanks. Feel free to merge.
